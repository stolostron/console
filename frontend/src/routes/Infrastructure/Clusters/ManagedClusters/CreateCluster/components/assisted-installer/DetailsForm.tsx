/* Copyright Contributors to the Open Cluster Management project */
/* eslint-disable react-hooks/exhaustive-deps */
import { useCallback, useRef, useEffect, useState, useMemo, Dispatch, SetStateAction } from 'react'
import { FormikProps } from 'formik'
import { CIM } from 'openshift-assisted-ui-lib'
import { set, get, isEqual, startCase, camelCase, debounce } from 'lodash'
// eslint-disable-next-line
import { TFunction } from 'react-i18next'
import { SelectOption, Text } from '@patternfly/react-core'
import { Link } from 'react-router-dom'
import { NavigationPath } from '../../../../../../../NavigationPath'
import { Secret, ManagedClusterSet } from '../../../../../../../resources'
import { useCanJoinClusterSets, useMustJoinClusterSet } from '../../../../ClusterSets/components/useCanJoinClusterSets'
import { useClusterImages, getDefault } from './utils'
import { useSharedAtoms, useSharedRecoil, useRecoilState, useRecoilValue } from '../../../../../../../shared-recoil'

import { getValue } from '../../../../../../../components/TemplateEditor'
import { AcmLabelsInput, AcmSelect } from '../../../../../../../ui-components'
import { useTranslation } from '../../../../../../../lib/acm-i18next'

const {
    ACMClusterDeploymentDetailsStep,
    FeatureGateContextProvider,
    ACMFeatureSupportLevelProvider,
    ACM_ENABLED_FEATURES,
    labelsToArray,
    LoadingState,
    getVersionFromReleaseImage,
} = CIM

type FormControl = {
    active: CIM.ClusterDetailsValues & {
        managedClusterSet?: string
        additionalLabels?: {
            [x: string]: string
        }[]
    }
    disabled?: VoidFunction
    reverse?: (control: { active: CIM.ClusterDetailsValues }, templateObject: any) => void
    validate?: VoidFunction
    summary?: VoidFunction
    step?: any
    additionalProps?: { [x: string]: string }
}

type DetailsFormProps = {
    control: FormControl
    handleChange: (control: FormControl) => void
    controlProps: Secret
}
const fields: any = {
    name: { path: 'ClusterDeployment[0].metadata.name' },
    baseDnsDomain: { path: 'ClusterDeployment[0].spec.baseDomain' },
    openshiftVersion: { path: 'AgentClusterInstall[0].spec.imageSetRef.name' },
    pullSecret: {},
}

export const getExtensionAfter = ({
    t,
    control,
    handleChange,
    canJoinClusterSets,
    managedClusterSet,
    setManagedClusterSet,
    mustJoinClusterSet,
    additionalLabels,
    setAdditionaLabels,
}: {
    t: TFunction
    control: FormControl
    handleChange: (control: FormControl) => void
    canJoinClusterSets: ManagedClusterSet[] | undefined
    managedClusterSet: string | undefined
    setManagedClusterSet: Dispatch<SetStateAction<string | undefined>>
    mustJoinClusterSet: boolean | undefined
    additionalLabels: Record<string, string> | undefined
    setAdditionaLabels: Dispatch<SetStateAction<Record<string, string> | undefined>>
}) => ({
    // the "key" references element preceeding the one which is being added
    name: (
        <AcmSelect
            id="managedClusterSet"
            label={t('import.form.managedClusterSet.label')}
            placeholder={
                canJoinClusterSets?.length === 0
                    ? t('import.no.cluster.sets.available')
                    : t('import.form.managedClusterSet.placeholder')
            }
            labelHelp={t('import.form.managedClusterSet.labelHelp')}
            value={managedClusterSet}
            onChange={(value) => {
                setManagedClusterSet(value)
                control.active = { ...control.active, managedClusterSet: value }
                handleChange(control)
            }}
            isDisabled={canJoinClusterSets === undefined || canJoinClusterSets.length === 0}
            hidden={canJoinClusterSets === undefined}
            helperText={
                <Text component="small">
                    <Link to={NavigationPath.clusterSets}>{t('import.manage.cluster.sets')}</Link>
                </Text>
            }
            isRequired={mustJoinClusterSet}
        >
            {canJoinClusterSets?.map((mcs) => (
                <SelectOption key={mcs.metadata.name} value={mcs.metadata.name}>
                    {mcs.metadata.name}
                </SelectOption>
            ))}
        </AcmSelect>
    ),
    openshiftVersion: (
        <AcmLabelsInput
            id="additionalLabels"
            label={t('import.form.labels.label')}
            buttonLabel={t('label.add')}
            value={additionalLabels}
            onChange={(label) => setAdditionaLabels(label)}
            placeholder={t('labels.edit.placeholder')}
            isDisabled={false}
        />
    ),
})

const DetailsForm: React.FC<DetailsFormProps> = ({ control, handleChange, controlProps }) => {
    const { clusterDeploymentsState, clusterImageSetsState } = useSharedAtoms()
    const { waitForAll } = useSharedRecoil()
    const [clusterDeployments] = useRecoilState(clusterDeploymentsState)
    const [clusterImageSets] = useRecoilValue(waitForAll([clusterImageSetsState]))
    const formRef = useRef<FormikProps<any>>(null)
    const { t } = useTranslation()

    const { canJoinClusterSets } = useCanJoinClusterSets()
    const mustJoinClusterSet = useMustJoinClusterSet()
    const [managedClusterSet, setManagedClusterSet] = useState<string | undefined>()
    const [additionalLabels, setAdditionaLabels] = useState<Record<string, string> | undefined>({})

    const getVersion = (versionName = '') => {
        const clusterImage = clusterImageSets.find((clusterImageSet) => clusterImageSet.metadata?.name == versionName)
        return getVersionFromReleaseImage(clusterImage?.spec?.releaseImage) || versionName
    }

    useEffect(() => {
        if (control.disabled && formRef?.current) {
            Array.from(document.forms[0].elements as HTMLCollectionOf<HTMLElement>).forEach((item: HTMLElement) => {
                item.setAttribute('disabled', 'true')
            })
        }
    }, [control.disabled, formRef.current])

    useEffect(() => {
        if (formRef?.current && control.active && control.active !== formRef?.current?.values) {
            formRef?.current?.setValues(control.active, true)
        }

        control.reverse = (
            control: {
                active: FormControl['active']
            },
            templateObject: any
        ) => {
            const active = {
                ...control.active,
                managedClusterSet: control.active.managedClusterSet,
                additionalLabels: control.active.additionalLabels,
            }
            if (control.active.managedClusterSet !== managedClusterSet) {
                setManagedClusterSet(control.active.managedClusterSet)
            }
            Object.keys(fields).forEach((key) => {
                const path = fields[key].path
                if (path) {
                    set(active, key, getValue(templateObject, path) || '')
                }
            })
            if (!isEqual(active, control.active)) {
                control.active = active
            }

            if (formRef.current && !isEqual(active, formRef.current.values)) {
                formRef.current.setValues(active)
            }
        }
        control.validate = () => {
            return formRef?.current?.submitForm().then(() => {
                return formRef?.current?.errors
            })
        }
        control.summary = () => {
            return Object.keys(fields).map((key) => {
                let desc = get(control, `active.${key}`)
                if (key === 'openshiftVersion') {
                    desc = getVersion(get(control, `active.${key}`))
                }
                return {
                    term: startCase(camelCase(key)),
                    desc: desc,
                    exception: get(control, `errors.${key}`),
                }
            })
        }
    }, [control])

    const clusterImages = useClusterImages()

    const usedClusterNames = useMemo(() => clusterDeployments.map((cd) => cd.metadata.name || ''), [])

    const extensionAfter = getExtensionAfter({
        t,
        control,
        handleChange,
        canJoinClusterSets,
        managedClusterSet,
        setManagedClusterSet,
        mustJoinClusterSet,
        additionalLabels,
        setAdditionaLabels,
    })

    useEffect(() => {
        control.active = { ...control.active, managedClusterSet }
        handleChange(control)
    }, [managedClusterSet])

    const onValuesChanged = useCallback(
        debounce((formikValues, initRender) => {
            const values = {
                ...formikValues,
                managedClusterSet: control.active.managedClusterSet,
                additionalLabels: control.active.additionalLabels,
            }
            if (!isEqual(values, control.active)) {
                if (!initRender || control.active.name === '') {
                    control.active = values
                }
                control.step.title.isComplete = false
                handleChange(control)
            }
        }),
        []
    )

    useEffect(() => {
        control.active = {
            ...control.active,
            additionalLabels: labelsToArray(additionalLabels).map((keyValue) => {
                const [key, value] = keyValue.split('=', 2)
                return { key, value }
            }),
        }
        handleChange(control)
    }, [additionalLabels])

    useEffect(() => {
        control.active = {
            ...control.active,
            pullSecret: getDefault([controlProps?.stringData?.pullSecret, control.active.pullSecret, '']),
            baseDnsDomain: getDefault([controlProps?.stringData?.baseDomain, control.active.baseDnsDomain, '']),
        }
        handleChange(control)
    }, [controlProps?.metadata.uid, controlProps?.stringData?.pullSecret, controlProps?.stringData?.baseDomain])

    return clusterImages ? (
        <FeatureGateContextProvider features={ACM_ENABLED_FEATURES}>
            <ACMFeatureSupportLevelProvider clusterImages={clusterImages as CIM.ClusterImageSetK8sResource[]}>
                <ACMClusterDeploymentDetailsStep
                    formRef={formRef}
                    onValuesChanged={onValuesChanged}
                    clusterImages={clusterImages as CIM.ClusterImageSetK8sResource[]}
                    usedClusterNames={usedClusterNames}
                    extensionAfter={extensionAfter}
                />
            </ACMFeatureSupportLevelProvider>
        </FeatureGateContextProvider>
    ) : (
        <LoadingState />
    )
}

export default DetailsForm
