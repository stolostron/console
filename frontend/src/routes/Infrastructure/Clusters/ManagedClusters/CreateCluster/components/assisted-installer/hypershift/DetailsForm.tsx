/* Copyright Contributors to the Open Cluster Management project */
/* eslint-disable react-hooks/exhaustive-deps */
import { useCallback, useRef, useEffect, useState, useMemo, useContext } from 'react'
import { FormikProps } from 'formik'
import { CIM } from 'openshift-assisted-ui-lib'
import { set, get, isEqual, startCase, camelCase } from 'lodash'
import { getValue } from '../../../../../../../../components/TemplateEditor'
import { AcmLabelsInput, AcmSelect } from '../../../../../../../../ui-components'
import { useTranslation } from '../../../../../../../../lib/acm-i18next'
import { SelectOption, Text } from '@patternfly/react-core'
import { Link } from 'react-router-dom'
import {
    useCanJoinClusterSets,
    useMustJoinClusterSet,
} from '../../../../../ClusterSets/components/useCanJoinClusterSets'
import { useClusterImages } from '../utils'
import { Secret } from '../../../../../../../../resources'
import { HypershiftAgentContext } from './HypershiftAgentContext'
import { getClusterImageVersion } from './utils'
import { useSharedAtoms, useSharedRecoil, useRecoilValue } from '../../../../../../../../shared-recoil'
import { NavigationPath } from '../../../../../../../../NavigationPath'

const { HostedClusterDetailsStep, labelsToArray, LoadingState, getSupportedCM } = CIM

type FormControl = {
    active: CIM.ClusterDetailsValues & {
        managedClusterSet?: string
        additionalLabels?: {
            [x: string]: string
        }[]
        releaseImage?: string
    }
    disabled?: VoidFunction
    reverse?: (control: { active: CIM.ClusterDetailsValues }, templateObject: any) => void
    validate?: VoidFunction
    summary?: VoidFunction
    step?: any
    additionalProps?: { [x: string]: string }
}

type DetailsFormProps = {
    control: FormControl
    handleChange: (control: FormControl) => void
    controlProps: Secret
}
const fields: any = {
    name: { path: 'HostedCluster[0].metadata.name' },
    baseDnsDomain: { path: 'HostedCluster[0].spec.dns.baseDomain' },
    releaseImage: { path: 'HostedCluster[0].spec.release.image' },
    pullSecret: {},
}

const DetailsForm: React.FC<DetailsFormProps> = ({ control, handleChange, controlProps }) => {
    const { setClusterName, setReleaseImage } = useContext(HypershiftAgentContext)
    const { waitForAll } = useSharedRecoil()
    const { clusterDeploymentsState, clusterImageSetsState, configMapsState } = useSharedAtoms()
    const [clusterDeployments, clusterImageSets, configMaps] = useRecoilValue(
        waitForAll([clusterDeploymentsState, clusterImageSetsState, configMapsState])
    )
    const formRef = useRef<FormikProps<any>>(null)
    const { t } = useTranslation()

    const { canJoinClusterSets } = useCanJoinClusterSets()
    const mustJoinClusterSet = useMustJoinClusterSet()
    const [managedClusterSet, setManagedClusterSet] = useState<string | undefined>()
    const [additionalLabels, setAdditionaLabels] = useState<Record<string, string> | undefined>({})

    const supportedVersionCM = getSupportedCM(configMaps)

    useEffect(() => {
        if (formRef?.current && control.active && control.active !== formRef?.current?.values) {
            formRef?.current?.setValues(control.active, true)
        }

        control.reverse = (
            control: {
                active: FormControl['active']
            },
            templateObject: any
        ) => {
            const active = {
                ...control.active,
                managedClusterSet: control.active.managedClusterSet,
                additionalLabels: control.active.additionalLabels,
                openshiftVersion:
                    clusterImageSets.find((cis) => cis.spec?.releaseImage === control.active.releaseImage)?.metadata
                        .name || control.active.openshiftVersion,
            }
            if (control.active.managedClusterSet !== managedClusterSet) {
                setManagedClusterSet(control.active.managedClusterSet)
            }
            Object.keys(fields).forEach((key) => {
                const path = fields[key].path
                if (path) {
                    set(active, key, getValue(templateObject, path) || '')
                }
            })
            if (!isEqual(active, control.active)) {
                control.active = active
            }

            if (formRef.current && !isEqual(active, formRef.current.values)) {
                formRef.current.setValues(active)
            }
        }
        control.validate = () => {
            return formRef?.current?.submitForm().then(() => {
                return formRef?.current?.errors
            })
        }
        control.summary = () => {
            return Object.keys(fields).map((key) => {
                let desc = get(control, `active.${key}`)
                if (key === 'openshiftVersion') {
                    desc = getClusterImageVersion(get(control, `active.${key}`))
                }
                return {
                    term: startCase(camelCase(key)),
                    desc: desc,
                    exception: get(control, `errors.${key}`),
                }
            })
        }
    }, [control])

    const clusterImages = useClusterImages()

    const usedClusterNames = useMemo(() => clusterDeployments.map((cd) => cd.metadata.name || ''), [])

    const extensionAfter = {
        // the "key" references element preceeding the one which is being added
        name: (
            <AcmSelect
                id="managedClusterSet"
                label={t('import.form.managedClusterSet.label')}
                placeholder={
                    canJoinClusterSets?.length === 0
                        ? t('import.no.cluster.sets.available')
                        : t('import.form.managedClusterSet.placeholder')
                }
                labelHelp={t('import.form.managedClusterSet.labelHelp')}
                value={managedClusterSet}
                onChange={(value) => {
                    setManagedClusterSet(value)
                    control.active = { ...control.active, managedClusterSet: value }
                    handleChange(control)
                }}
                isDisabled={canJoinClusterSets === undefined || canJoinClusterSets.length === 0}
                hidden={canJoinClusterSets === undefined}
                helperText={
                    <Text component="small">
                        <Link to={NavigationPath.clusterSets}>{t('import.manage.cluster.sets')}</Link>
                    </Text>
                }
                isRequired={mustJoinClusterSet}
            >
                {canJoinClusterSets?.map((mcs) => (
                    <SelectOption key={mcs.metadata.name} value={mcs.metadata.name}>
                        {mcs.metadata.name}
                    </SelectOption>
                ))}
            </AcmSelect>
        ),
        openshiftVersion: (
            <AcmLabelsInput
                id="additionalLabels"
                label={t('import.form.labels.label')}
                buttonLabel={t('label.add')}
                value={additionalLabels}
                onChange={(label) => setAdditionaLabels(label)}
                placeholder={t('labels.edit.placeholder')}
                isDisabled={false}
            />
        ),
    }

    const onValuesChanged = useCallback((formikValues, initRender) => {
        setClusterName(formikValues.name)
        setReleaseImage(formikValues.openshiftVersion)
        const values = {
            ...formikValues,
            managedClusterSet: control.active.managedClusterSet,
            additionalLabels: control.active.additionalLabels,
            releaseImage: clusterImageSets.find(({ metadata }) => metadata.name === formikValues.openshiftVersion)?.spec
                ?.releaseImage,
        }
        if (!isEqual(values, control.active)) {
            if (!initRender || control.active.name === '') {
                control.active = values
            }
            control.step.title.isComplete = false
            handleChange(control)
        }
    }, [])

    useEffect(() => {
        control.active = {
            ...control.active,
            additionalLabels: labelsToArray(additionalLabels).map((keyValue) => {
                const [key, value] = keyValue.split('=', 2)
                return { key, value }
            }),
        }
        handleChange(control)
    }, [additionalLabels])

    useEffect(() => {
        control.active = {
            ...control.active,
            pullSecret: controlProps?.stringData?.pullSecret || '',
            baseDnsDomain: controlProps?.stringData?.baseDomain || '',
        }
        handleChange(control)
    }, [controlProps?.metadata.uid, controlProps?.stringData?.pullSecret, controlProps?.stringData?.baseDomain])

    return clusterImages ? (
        <HostedClusterDetailsStep
            formRef={formRef}
            onValuesChanged={onValuesChanged}
            clusterImages={clusterImages}
            usedClusterNames={usedClusterNames}
            extensionAfter={extensionAfter}
            supportedVersionsCM={supportedVersionCM}
        />
    ) : (
        <LoadingState />
    )
}

export default DetailsForm
