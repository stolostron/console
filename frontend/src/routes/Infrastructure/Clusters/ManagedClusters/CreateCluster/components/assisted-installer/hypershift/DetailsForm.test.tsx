/* Copyright Contributors to the Open Cluster Management project */
import { render } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom-v5-compat'
import { RecoilRoot } from 'recoil'
import { NavigationPath } from '../../../../../../../../NavigationPath'
import { nockIgnoreApiPaths, nockIgnoreRBAC, nockList } from '../../../../../../../../lib/nock-util'
import { waitForNocks, waitForText } from '../../../../../../../../lib/test-util'
import { clusterImageSet, mockClusterImageSet } from '../../../CreateCluster.sharedmocks'
import { IResource } from '../../../../../../../../resources'

import DetailsForm from './DetailsForm'

describe('DetailsForm', () => {
  const handleChange = jest.fn()
  const Component = () => {
    return (
      <RecoilRoot>
        <MemoryRouter initialEntries={[NavigationPath.createCluster]}>
          <Routes>
            <Route
              path={NavigationPath.createCluster}
              element={
                <DetailsForm
                  key={'key'}
                  control={{
                    active: {
                      name: '',
                      highAvailabilityMode: 'Full',
                      openshiftVersion: '',
                      pullSecret: '',
                      baseDnsDomain: '',
                      SNODisclaimer: false,
                      useRedHatDnsService: true,
                      enableDiskEncryptionOnMasters: false,
                      enableDiskEncryptionOnWorkers: false,
                      diskEncryptionMode: 'tpmv2',
                      diskEncryptionTangServers: [],
                      diskEncryption: {},
                      cpuArchitecture: '',
                      platform: 'baremetal',
                    },
                    step: {
                      title: {
                        isComplete: false,
                      },
                    },
                  }}
                  controlProps={{
                    apiVersion: 'v1',
                    kind: 'Secret',
                    metadata: {},
                  }}
                  handleChange={handleChange}
                />
              }
            />
          </Routes>
        </MemoryRouter>
      </RecoilRoot>
    )
  }

  test('it renders', async () => {
    nockIgnoreRBAC()
    nockIgnoreApiPaths()
    const initialNocks = [nockList(clusterImageSet as IResource, mockClusterImageSet as IResource[])]
    const { container } = render(<Component />)
    await waitForNocks(initialNocks)
    await waitForText('ai:Cluster name')
    expect(container).toMatchSnapshot()
  })
})
