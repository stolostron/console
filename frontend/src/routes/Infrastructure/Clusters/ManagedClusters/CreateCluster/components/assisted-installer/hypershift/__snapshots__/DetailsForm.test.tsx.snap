// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsForm it renders 1`] = `
<div>
  <form
    class="pf-v5-c-form"
    novalidate=""
  >
    <div
      class="pf-v5-c-form__group"
      id="form-control__form-input-name-field"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="form-input-name-field"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            ai:Cluster name
          </span>
          <span
            aria-hidden="true"
            class="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <div
          class="pf-v5-l-split"
        >
          <div
            class="pf-v5-l-split__item pf-m-fill"
          >
            <span
              class="pf-v5-c-form-control"
            >
              <input
                aria-describedby="form-input-name-field-helper"
                aria-invalid="false"
                data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                data-ouia-component-type="PF5/TextInput"
                data-ouia-safe="true"
                id="form-input-name-field"
                name="name"
                placeholder="ai:Enter cluster name"
                required=""
                type="text"
                value=""
              />
            </span>
          </div>
          <div
            class="pf-v5-l-split__item"
          />
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      id="managedClusterSet-label"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="managedClusterSet"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            Cluster set
          </span>
        </label>
         
        <div
          style="display: contents;"
        >
          <button
            aria-disabled="false"
            aria-label="More info"
            class="pf-v5-c-button pf-m-plain pf-v5-c-form__group-label-help"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            id="managedClusterSet-label-help-button"
            style="--pf-v5-c-form__group-label-help--TranslateY: 0;"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <div
          class="pf-v5-c-select"
          data-ouia-component-id="OUIA-Generated-Select-single-1"
          data-ouia-component-type="PF5/Select"
          data-ouia-safe="true"
        >
          <div
            class="pf-v5-c-select__toggle pf-m-disabled pf-m-typeahead"
          >
            <div
              class="pf-v5-c-select__toggle-wrapper"
            >
              <span
                class="pf-v5-c-select__toggle-text"
              >
                <span
                  style="color: rgb(102, 102, 102);"
                >
                  No cluster sets available
                </span>
              </span>
            </div>
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Options menu"
              aria-labelledby="managedClusterSet-label pf-select-toggle-id-1"
              class="pf-v5-c-button pf-v5-c-select__toggle-button pf-m-plain"
              disabled=""
              id="pf-select-toggle-id-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg pf-v5-c-select__toggle-arrow"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="pf-v5-c-form__helper-text"
          id="managedClusterSet-helper"
        >
          <div
            class="pf-v5-l-split"
          >
            <div
              class="pf-v5-l-split__item pf-m-fill"
            >
              <div
                class="pf-v5-c-helper-text"
              >
                <div
                  class="pf-v5-c-helper-text__item"
                >
                  <span
                    class="pf-v5-c-helper-text__item-text"
                  >
                    <small
                      class=""
                      data-ouia-component-id="OUIA-Generated-Text-1"
                      data-ouia-component-type="PF5/Text"
                      data-ouia-safe="true"
                      data-pf-content="true"
                    >
                      <a
                        href="/multicloud/infrastructure/clusters/sets"
                      >
                        Manage cluster sets
                      </a>
                    </small>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      id="form-control__form-input-baseDnsDomain-field"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="form-input-baseDnsDomain-field"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            ai:Base domain
          </span>
          <span
            aria-hidden="true"
            class="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <div
          class="pf-v5-l-split"
        >
          <div
            class="pf-v5-l-split__item pf-m-fill"
          >
            <span
              class="pf-v5-c-form-control"
            >
              <input
                aria-describedby="form-input-baseDnsDomain-field-helper"
                aria-invalid="false"
                data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                data-ouia-component-type="PF5/TextInput"
                data-ouia-safe="true"
                id="form-input-baseDnsDomain-field"
                name="baseDnsDomain"
                placeholder="example.com"
                required=""
                type="text"
                value=""
              />
            </span>
          </div>
          <div
            class="pf-v5-l-split__item"
          />
        </div>
        <div
          class="pf-v5-c-form__helper-text"
        >
          <div
            class="pf-v5-c-helper-text"
          >
            <div
              class="pf-v5-c-helper-text__item pf-m-error"
              id="form-input-baseDnsDomain-field-helper"
            >
              <span
                class="pf-v5-c-helper-text__item-text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      id="form-control__form-input-openshiftVersion-field"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="form-input-openshiftVersion-field"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            ai:OpenShift version
          </span>
          <span
            aria-hidden="true"
            class="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <span
          class="pf-v5-c-form-control"
        >
          <select
            aria-describedby="form-input-openshiftVersion-field-helper"
            aria-invalid="false"
            data-ouia-component-id="OUIA-Generated-FormSelect-default-1"
            data-ouia-component-type="PF5/FormSelect"
            data-ouia-safe="true"
            id="form-input-openshiftVersion-field"
            name="openshiftVersion"
            required=""
          >
            <option
              class=""
              value="ocp-release48"
            >
              OpenShift 4.8.15
            </option>
          </select>
          <span
            class="pf-v5-c-form-control__utilities"
          >
            <span
              class="pf-v5-c-form-control__toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      id="form-control__form-input-pullSecret-field"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="form-input-pullSecret-field"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            ai:Pull secret
          </span>
          <span
            aria-hidden="true"
            class="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
        <div
          style="display: contents;"
        >
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-plain pf-v5-c-form__group-label-help pf-v5-u-p-0"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            <span
              class="pf-v5-c-icon pf-m-inline"
            >
              <span
                class="pf-v5-c-icon__content"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <span
          class="pf-v5-c-form-control pf-m-resize-both"
        >
          <textarea
            aria-describedby="form-input-pullSecret-field-helper"
            aria-invalid="false"
            id="form-input-pullSecret-field"
            name="pullSecret"
            required=""
            rows="10"
            style="resize: vertical;"
          />
        </span>
        <div
          class="pf-v5-c-form__helper-text"
        >
          <div
            class="pf-v5-c-helper-text"
          >
            <div
              class="pf-v5-c-helper-text__item"
              id="form-input-pullSecret-field-helper"
            >
              <span
                class="pf-v5-c-helper-text__item-text"
              >
                ai:A Red Hat account's pull secret can be found in 
                 
                <a
                  href="https://console.redhat.com/openshift/install/pull-secret"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  ai:OpenShift Cluster Manager
                   
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                    />
                  </svg>
                </a>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
